src/conf/order.rs

```diff
@@ -289,17 +289,94 @@ fn parse_ip_identifier(
     Ok(Some(out))
 }
 
-/// Checks if the value is a valid domain name and returns a canonical (lowercase) form,
-/// reallocated on the configuration pool if necessary.
-fn validate_host(cf: &ngx_conf_t, host: &'static str) -> Result<&'static str, Status> {
-    let mut host = ngx_str_t {
-        data: host.as_ptr().cast_mut(),
-        len: host.len(),
-    };
-    let rc = Status(unsafe { nginx_sys::ngx_http_validate_host(&mut host, cf.pool, 0) });
-    if rc != Status::NGX_OK {
-        return Err(rc);
+#[derive(Debug, Error, PartialEq)]
+pub enum NameError {
+    #[error("invalid character {0:x} at position {1}")]
+    Invalid(u8, usize),
+    #[error("unexpected character '{0}' at position {1}")]
+    Unexpected(char, usize),
+    #[error("does not end with a label")]
+    NotALabel,
+}
+
+/// Validates DNS name according to RFC 1123 + '_' + leading dot and wildcard.
+///
+/// Returns true if the name needs to be reallocated and converted to lower case.
+///
+/// Both the ACME protocol and the NGINX configuration require ASCII-encoded domain names.
+/// See: https://nginx.org/en/docs/http/server_names.html
+fn validate_dns_identifier(name: &str) -> Result<bool, NameError> {
+    #[derive(PartialEq, Eq)]
+    enum State {
+        Start,
+        Label,
+        Dot,
+        Wildcard,
     }
 
-    unsafe { super::conf_value_to_str(&host) }.map_err(|_| Status::NGX_ERROR)
+    let mut alloc = false;
+    let mut state = State::Start;
+
+    for (i, ch) in name.bytes().enumerate() {
+        state = match ch {
+            0x00..=0x20 | 0x7f.. => return Err(NameError::Invalid(ch, i)),
+
+            b'*' if state == State::Start => State::Wildcard,
+            b'.' if state != State::Dot => State::Dot,
+            b'/' | b':' => return Err(NameError::Unexpected(ch as _, i)),
+
+            // '*' can only be followed by '.'.
+            _ if state == State::Wildcard => return Err(NameError::Unexpected(ch as _, i)),
+
+            b'A'..=b'Z' => {
+                alloc = true;
+                State::Label
+            }
+            b'0'..=b'9' | b'a'..=b'z' | b'-' | b'_' => State::Label,
+
+            _ => return Err(NameError::Unexpected(ch as _, i)),
```

Comment:
`validate_dns_identifier ` accepts "-example" or "example-"

RFC 1123 references RFC952 which does not allow "-" at the ends: 
```
 B. Lexical grammar
<name>  ::= <let>[*[<let-or-digit-or-hyphen>]<let-or-digit>]
```